File: cpp_in_python_example/test_model.py
#!/usr/bin/env python
from math import *
import matplotlib.pyplot as plt
from helper_functions import HEADER_LENGTH, HEADER_WIDTH, TRAILER_LENGTH, TRAILER_WIDTH, HL_FRONT, TL_BACK



class Point:
    def __init__(self, x , y):
        self.x = x
        self.y = y

class Vector:
    def __init__(self, x , y):
        self.values = (x,y)

    def magnitude(self):
        #magnitude of the vector
        (x,y)= self.values
        return math.sqrt(x*x + y*y)

    def rotate(self, theta):
        #rotate the vector theta degrees
        theta = math.radians(theta)
        dc, ds = math.cos(theta), math.sin(theta)
        (x, y) = self.values
        (x, y) = dc*x - ds*y, ds*x + dc*y
        return Vector(x, y)

point = Point(0,0)
point1 = Point(0,0)



class truck:

    def calculateCorners(self, pointFront, th1, th2):

        point = self.back_middle_trailer(pointFront, th1, th2)

        # hitch joint
        x2 = point.x + (TRAILER_LENGTH + TL_BACK) *cos(th2);
        y2 = point.y + (TRAILER_LENGTH + TL_BACK) *sin(th2);

        #print x2, y2


        #left back
        x4 = point.x - cos(pi/2-th2)*TRAILER_WIDTH/2;
        y4 = point.y + sin(pi/2-th2)*TRAILER_WIDTH/2;
        print "left back:", x4, y4

        #right back
        x5 = point.x + cos(pi/2-th2)*TRAILER_WIDTH/2;
        y5 = point.y - sin(pi/2-th2)*TRAILER_WIDTH/2;
        print "right back:", x5, y5


        #left back axis
        x12 = x4 + TL_BACK * cos(th2)
        y12 = y4 + TL_BACK * sin(th2)
        print "left back axis:", x12, y12

        #right back axis
        x13 = x5 + TL_BACK * cos(th2)
        y13 = y5 + TL_BACK * sin(th2)
        print "right back axis:", x13, y13



        #left joint wheel
        x6 = x2 - cos(pi/2-th1)*HEADER_WIDTH/2;
        y6 = y2 + sin(pi/2-th1)*HEADER_WIDTH/2;

        #right joint wheel
        x7 = x2 + cos(pi/2-th1)*HEADER_WIDTH/2;
        y7 = y2 - sin(pi/2-th1)*HEADER_WIDTH/2;

        #left front axis
        x8 = x6 + (HEADER_LENGTH-5.0)*cos(th1);
        y8 = y6 + (HEADER_LENGTH-5.0)*sin(th1);
        print "left front axis:", x8, y8

        #right front axis
        x9 = x7 + (HEADER_LENGTH-5.0)*cos(th1);
        y9 = y7 + (HEADER_LENGTH-5.0)*sin(th1);
        print "right front axis:", x9, y9


        #left front

        x10 = x6 + ((HEADER_LENGTH-5.0) + HL_FRONT)*cos(th1);
        y10 = y6 + ((HEADER_LENGTH-5.0) + HL_FRONT)*sin(th1);
        print "left front:", x10, y10

        #right front

        x11 = x7 + ((HEADER_LENGTH-5.0) + HL_FRONT)*cos(th1);
        y11 = y7 + ((HEADER_LENGTH-5.0) + HL_FRONT)*sin(th1);
        print "right front:", x11, y11



        return((x8,y8), (x9,y9), (x6,y6), (x7,y7), (x12, y12), (x13, y13), (x10, y10), (x11, y11), (x4, y4), (x5, y5))

    def rightbackWheel(self):

        point = self.back_middle_trailer()

        x5 = point.x + cos(pi/2-th2)*TRAILER_WIDTH/2;
        y5 = point.y - sin(pi/2-th2)*TRAILER_WIDTH/2;

    def back_middle_trailer(self, pointFront, th1, th2):
        jpx = pointFront.x - cos(th1) * (HEADER_LENGTH-5.0)
        jpy = pointFront.y - sin(th1) * (HEADER_LENGTH-5.0)

        px = jpx - cos(th2) * (TRAILER_LENGTH + TL_BACK)
        py = jpy - sin(th2) * (TRAILER_LENGTH + TL_BACK)
        return Point(px,py)




if __name__ == '__main__':

    truck = truck()
    truck.calculateCorners(Point(200, 200), radians(-90), radians(-90))

File: scripts/test_model.cpp
#include "model.hpp"
#include "helper_functions.hpp"
#include <math.h>
#include <iostream>


Truck::Truck() {
    left_front = new Point(0.0, 0.0);
    right_front = new Point(0.0, 0.0);
    left_back = new Point(0.0, 0.0);
    right_back = new Point(0.0, 0.0);

    left_front_wheel = new Point(0.0, 0.0);
    right_front_wheel = new Point(0.0, 0.0);
    left_back_wheel = new Point(0.0, 0.0);
    right_back_wheel = new Point(0.0, 0.0);
}


Truck::~Truck() {
    delete left_front;
    delete right_front;
    delete left_back;
    delete right_back;

    delete left_front_wheel;
    delete right_front_wheel;
    delete left_back_wheel;
    delete right_back_wheel;
}


void Truck::setCorners(Point *pointFront, double th1, double th2) {
    //std::cout << "theta1: "<< th1 << " cos: " << cos(th1) << " sin: " << sin(th1) << std::endl;
    //std::cout << "theta2: "<< th2 << " cos: " << cos(th2) << " sin: " << sin(th2) << std::endl << std::endl;

    Point *point = backMiddleTrailer(pointFront, th1, th2);

    // Hitch joint
    double x2 = point->x + (TRAILER_LENGTH + TL_BACK) * cos(th2);
    double y2 = point->y + (TRAILER_LENGTH + TL_BACK) * sin(th2);

    // Left back
    double x4 = point->x - cos(M_PI/2 - th2) * TRAILER_WIDTH/2;
    double y4 = point->y + sin(M_PI/2 - th2) * TRAILER_WIDTH/2;
    left_back->setValues(x4, y4);
    std::cout << "left back: " << x4 << " " << y4 << std::endl;

    // Right back
    double x5 = point->x + cos(M_PI/2 - th2) * TRAILER_WIDTH/2;
    double y5 = point->y - sin(M_PI/2 - th2) * TRAILER_WIDTH/2;
    right_back->setValues(x5, y5);
    std::cout << "right back: " << x5 << " " << y5 << std::endl;

    // Left back axis
    double x12 = x4 + TL_BACK * cos(th2);
    double y12 = y4 + TL_BACK * sin(th2);
    left_back_wheel->setValues(x12, y12);
    //std::cout << std::endl << x4 << " " << TL_BACK << " " << cos(th2) << " ekv: " << tl_back * cos(th2) << std::endl;
    //std::cout << y4 << " " << TL_BACK << " " << sin(th2) << " ekv: " << tl_back * sin(th2) << std::endl << std::endl;
    std::cout << "left back axis: " << x12 << " " << y12 << std::endl;

    // Right back axis
    double x13 = x5 + (TL_BACK * cos(th2));
    double y13 = y5 + (TL_BACK * sin(th2));
    right_back_wheel->setValues(x13, y13);
    std::cout << "right back axis: " << x13 << " " << y13 << std::endl;

    // Left joint wheel
    double x6 = x2 - cos(M_PI/2 - th1) * HEADER_WIDTH/2;
    double y6 = y2 + sin(M_PI/2 - th1) * HEADER_WIDTH/2;

    // Right joint wheel
    double x7 = x2 + cos(M_PI/2 - th1) * HEADER_WIDTH/2;
    double y7 = y2 - sin(M_PI/2 - th1) * HEADER_WIDTH/2;

    // Left front axis
    double x8 = x6 + (HEADER_LENGTH-5.0) * cos(th1);
    double y8 = y6 + (HEADER_LENGTH-5.0) * sin(th1);
    left_front_wheel->setValues(x8, y8);
    //std::cout << "theta1: " << th1 << " " << cos(th1) << " " << sin(th1) << std::endl;
    std::cout << "left front axis: " << x8 << " " << y8 << std::endl;

    // Right front axis
    double x9 = x7 + (HEADER_LENGTH-5.0) * cos(th1);
    double y9 = y7 + (HEADER_LENGTH-5.0) * sin(th1);
    right_front_wheel->setValues(x9, y9);
    std::cout << "right front axis: " << x9 << " " << y9 << std::endl;

    // Left front
    double x10 = x6 + ((HEADER_LENGTH-5.0) + HL_FRONT) * cos(th1);
    double y10 = y6 + ((HEADER_LENGTH-5.0) + HL_FRONT) * sin(th1);
    left_front->setValues(x10, y10);
    std::cout << "left front: " << x10 << " " << y10 << std::endl;

    // Right front
    double x11 = x7 + ((HEADER_LENGTH-5.0) + HL_FRONT) * cos(th1);
    double y11 = y7 + ((HEADER_LENGTH-5.0) + HL_FRONT) * sin(th1);
    right_front->setValues(x11, y11);
    std::cout << "right front: " << x11 << " " << y11 << std::endl;

    //std::cout << std::endl << "theta1: "<< th1 << " cos: " << cos(th1) << " sin: " << sin(th1) << std::endl;
    //std::cout << "theta2: "<< th2 << " cos: " << cos(th2) << " sin: " << sin(th2) << std::endl;
}


Point * Truck::backMiddleTrailer(Point *pointFront, double th1, double th2) {
    double jpx = pointFront->x - cos(th1) * (HEADER_LENGTH-5.0);
    double jpy = pointFront->y - sin(th1) * (HEADER_LENGTH-5.0);

    double px = jpx - cos(th2) * (TRAILER_LENGTH + TL_BACK);
    double py = jpy - sin(th2) * (TRAILER_LENGTH + TL_BACK);

    return new Point(px, py);
}



int main() {
    Truck *truck = new Truck();
    truck->setCorners(new Point(200, 200), radians(-90), radians(-90));
    //truck->setCorners(new Point(200, 200), radians(0), radians(0));
}


/*

g++ -std=gnu++11 test_model.cpp pathPlanning.cpp Point.cpp track_checker.cpp vehicleState.cpp vehicleState_error.cpp helper_functions.cpp error_calc.cpp

*/